### PR TITLE
add retry limit for excluded backoff type to avoid infinite retry

### PR DESCRIPTION
**Pull request by cfzjywxk**
Ref: https://github.com/tikv/client-go/issues/1001

Try to add retry limit for excluded backoff type to avoid infinite retry.

### CODE REVIEW

**Review by zyguan — APPROVED** [review]

File: internal/retry/backoff.go [hunk 1]
```diff
@@ -143,7 +143,13 @@ func (b *Backoffer) BackoffWithCfgAndMaxSleep(cfg *Config, maxSleepMs int, err e
 	if b.noop {
 		return err
 	}
-	if b.maxSleep > 0 && (b.totalSleep-b.excludedSleep) >= b.maxSleep {
+	maxBackoffTimeExceeded := (b.totalSleep - b.excludedSleep) >= b.maxSleep
+	maxExcludedTimeExceeded := false
+	if maxLimit, ok := isSleepExcluded[cfg.name]; ok {
+		maxExcludedTimeExceeded = b.excludedSleep >= maxLimit && b.excludedSleep >= b.maxSleep
```
Comment (ekexium) [inline internal/retry/backoff.go:149]: From the name `maxLimit`, I would suppose the latter part is not necessary. In practice it won't make big difference though.

**Review by ekexium — APPROVED** [review]

File: internal/retry/backoff.go [hunk 2]
```diff
@@ -163,7 +169,8 @@ func (b *Backoffer) BackoffWithCfgAndMaxSleep(cfg *Config, maxSleepMs int, err e
 			backoffDetail.WriteString(":")
 			backoffDetail.WriteString(strconv.Itoa(times))
 		}
-		errMsg += fmt.Sprintf("\ntotal-backoff-times: %v, backoff-detail: %v", totalTimes, backoffDetail.String())
+		errMsg += fmt.Sprintf("\ntotal-backoff-times: %v, backoff-detail: %v, maxBackoffTimeExceeded: %v, maxBackoffTimeExceeded: %v",
```
Comment (crazycs520) [inline internal/retry/backoff.go:172]:
```suggestion
		errMsg += fmt.Sprintf("\ntotal-backoff-times: %v, backoff-detail: %v, maxBackoffTimeExceeded: %v, maxExcludedTimeExceeded: %v",
```

File: internal/retry/backoff.go (hunk 1 above)
Comment (cfzjywxk) [reply internal/retry/backoff.go:149]: In normal circumstances, this check at the end is not necessary. It's mainly to guard against some erroneous inputs or configurations.

File: internal/retry/config.go [hunk 3]
```diff
@@ -130,11 +130,18 @@ var (
 	BoTxnLockFast = NewConfig(txnLockFastName, &metrics.BackoffHistogramLockFast, NewBackoffFnCfg(2, 3000, EqualJitter), tikverr.ErrResolveLockTimeout)
 )
 
-var isSleepExcluded = map[string]struct{}{
-	BoTiKVServerBusy.name: {},
+var isSleepExcluded = map[string]int{
+	BoTiKVServerBusy.name: 600000, // The max excluded limit is 10min.
```
Comment (jackysp) [inline internal/retry/config.go:134]: It's a bit too far to add a new limit of it, I'm afraid that someday it will be required to be configurable.

Comment (cfzjywxk) [reply internal/retry/config.go:134]: This could only be used in some extreme cases like the service is unavailable but the leader is still valid.

**Pull request merged**